fix(agw): Modifed code to port hash list of ue contexts to protobuf map

rsarwad opened the pull request:
## Summary
fix(agw): Modifed code to port hash list of ue contexts to protobuf map
## Test Plan
Executed s1ap sanity test suite and unit test cases.
Because of DCO check failure,  the changes have been duplicated PR, https://github.com/magma/magma/pull/14342

rsarwad committed:
Modified code to fix compilation errors

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Removed files and structures that are no more required

rsarwad committed:
Ported heid hash list protobuf map

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into 11190_spgw_teid_hl

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Resolved merge conflicts

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Addressed self comments

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
fix(agw): Modiifed code to port hash list of ue contexts to protobuf map

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
fix(agw): Modiifed code to port hash list of ue contexts to protobuf map

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into 11190_spgw_ue_context_ht

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

pruthvihebbani commented on lte/gateway/c/core/oai/include/spgw_state.hpp:
```diff
@@ -44,9 +44,9 @@ void put_spgw_state(void);
  * Returns pointer to SPGW UE state
  * @return hashtable_ts_t struct with SPGW UE context structs as data
```
Update the return type in comment

pruthvihebbani commented on lte/gateway/c/core/oai/tasks/sgw/sgw.hpp:
```diff
@@ -1,70 +0,0 @@
-/*
```
Is this not used anywhere?

rsarwad replied:
sgw.hpp is nowhere used

ssanadhya commented on lte/gateway/c/core/oai/tasks/sgw/sgw_context_manager.cpp:
```diff
@@ -157,35 +157,16 @@ sgw_cm_create_bearer_context_information_in_collection(teid_t teid) {
-  }
-
   /*
    * Trying to insert the new tunnel into the tree.
```
Please update the comment

```suggestion
   * Trying to insert the new tunnel into the map.
```

ssanadhya commented on lte/gateway/c/core/oai/tasks/sgw/sgw_context_manager.cpp:
```diff
@@ -196,36 +177,42 @@ sgw_cm_create_bearer_context_information_in_collection(teid_t teid) {
-                    "Failed to free teid from state_teid_ht \n");
-    return temp;
+                    "Failed to free teid from state_teid_map \n");
+    return magma::PROTO_MAP_KEY_ALREADY_EXISTS;
```
This should be an explicit error, not that key already exists. Can we add an error code for `PROTO_MAP_REMOVE_KEY_FAILED`

rsarwad replied:
The above comment is addressed at https://github.com/magma/magma/pull/13895/files#diff-8dbb5820f4aef5a4b6295c1040d13d405728e409346d63b37cfac400b8004312R189

ssanadhya commented on lte/gateway/c/core/oai/tasks/sgw/sgw_context_manager.cpp:
```diff
@@ -318,35 +305,44 @@ sgw_eps_bearer_ctxt_t* sgw_cm_get_eps_bearer_entry(
+  state_teid_map_t* state_teid_map = get_spgw_teid_state();
+  if (!state_teid_map) {
+    OAILOG_ERROR(LOG_SPGW_APP, "Failed to get state_teid_map");
+    return spgw_bearer_context_info;
```
Better to just return a `nullptr` (which is the value of spgw_bearer_context_info at this point) since this is an error scenario.
```suggestion
    return nullptr;
```

ssanadhya commented on lte/gateway/c/core/oai/tasks/sgw/sgw_context_manager.cpp:
```diff
@@ -318,35 +305,44 @@ sgw_eps_bearer_ctxt_t* sgw_cm_get_eps_bearer_entry(
+  map_uint64_spgw_ue_context_t* state_ue_map = get_spgw_ue_state();
+  if (!state_ue_map) {
+    OAILOG_ERROR(LOG_SPGW_APP, "Failed to find spgw_ue_state");
+    OAILOG_FUNC_RETURN(LOG_SPGW_APP, ue_context_p);
```
Same comment as above
```suggestion
    OAILOG_FUNC_RETURN(LOG_SPGW_APP, nullptr);
```

ssanadhya commented on lte/gateway/c/core/oai/tasks/sgw/sgw_context_manager.cpp:
```diff
@@ -318,35 +305,44 @@ sgw_eps_bearer_ctxt_t* sgw_cm_get_eps_bearer_entry(
 
 spgw_ue_context_t* spgw_create_or_get_ue_context(imsi64_t imsi64) {
   OAILOG_FUNC_IN(LOG_SPGW_APP);
   spgw_ue_context_t* ue_context_p = NULL;
```
```suggestion
  spgw_ue_context_t* ue_context_p = nullptr;
```

ssanadhya commented on lte/gateway/c/core/oai/tasks/sgw/sgw_context_manager.cpp:
```diff
@@ -318,35 +305,44 @@ sgw_eps_bearer_ctxt_t* sgw_cm_get_eps_bearer_entry(
+  map_uint64_spgw_ue_context_t* state_ue_map = get_spgw_ue_state();
+  if (!state_ue_map) {
+    OAILOG_ERROR(LOG_SPGW_APP, "Failed to find spgw_ue_state");
+    OAILOG_FUNC_RETURN(LOG_SPGW_APP, ue_context_p);
```
```suggestion
    OAILOG_FUNC_RETURN(LOG_SPGW_APP, nullptr);
```

ssanadhya commented on lte/gateway/c/core/oai/tasks/sgw/sgw_handlers.cpp:
```diff
@@ -1240,10 +1240,14 @@ void sgw_handle_release_access_bearers_request(
 
   spgw_ue_context_t* ue_context_p = NULL;
```
Please replace NULL with nullptr for *.cpp files.

ssanadhya commented on lte/gateway/c/core/oai/tasks/sgw/sgw_handlers.cpp:
```diff
@@ -1636,8 +1640,9 @@ status_code_e sgw_handle_nw_initiated_actv_bearer_rsp(
     LIST_REMOVE(base_proc1, entries);
-    free_wrapper((void**)&spgw_context->sgw_eps_bearer_context_information
-                     .pending_procedures);
+    delete spgw_context->sgw_eps_bearer_context_information.pending_procedures;
```
Please use `free_cpp_wrapper`

ssanadhya commented on lte/gateway/c/core/oai/tasks/sgw/sgw_handlers.cpp:
```diff
@@ -1984,7 +1989,8 @@ void handle_failed_create_bearer_response(
             LIST_FIRST(sgw_context_p->pending_procedures);
         LIST_REMOVE(base_proc1, entries);
-        free_wrapper((void**)&sgw_context_p->pending_procedures);
+        delete sgw_context_p->pending_procedures;
```
Please use `free_cpp_wrapper`

ssanadhya commented on lte/gateway/c/core/oai/tasks/sgw/spgw_state.cpp:
```diff
@@ -81,21 +85,20 @@ void delete_spgw_ue_state(imsi64_t imsi64) {
+      delete_pending_procedures(
+          &(context_p->sgw_eps_bearer_context_information));
+
+      free_cpp_wrapper((void**)ptr);
```
Please fix lint warnings

pruthvihebbani replied:
Fix the warning

ssanadhya commented on lte/gateway/c/core/oai/tasks/sgw/spgw_state_converter.cpp:
```diff
@@ -215,8 +215,10 @@ void SpgwStateConverter::sgw_pdn_connection_to_proto(
     const oai::SgwPdnConnection& proto, sgw_pdn_connection_t* state_pdn) {
-  state_pdn->apn_in_use =
-      strndup(proto.apn_in_use().c_str(), proto.apn_in_use().length());
+  if ((proto.apn_in_use().length())) {
```
Can we use `proto.has_apn_in_use()` here?

rsarwad replied:
apn_in_use is of string type. For string data type, protobuf doesn't generate api, has_<field_name>;
It's in https://developers.google.com/protocol-buffers/docs/reference/cpp-generated#proto3_string
If field is not set then it returns empty string, but again in code we need to check empty string using memcmp; So I thought to use length of string.

ssanadhya commented on lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_handlers.cpp:
```diff
@@ -97,7 +97,7 @@ void sgw_remove_sgw_bearer_context_information(sgw_state_t* sgw_state,
       if (p1->sgw_s11_teid == teid) {
         LIST_REMOVE(p1, entries);
-        free_wrapper((void**)&p1);
+        free_cpp_wrapper((void**)&p1);
```
Please fix lint warning

pruthvihebbani replied:
Fix cpplint warning

ssanadhya commented on lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:
```diff
@@ -119,6 +118,7 @@ TEST(SPGWStateConverterTest, TestUEContextConversion) {
     EXPECT_TRUE(!memcmp(&want_sgw.last_known_cell_Id,
                         &got_sgw.last_known_cell_Id,
                         sizeof(want_sgw.last_known_cell_Id)));
+    sgw_free_ue_context((void**)&final_ctx);
```
Please fix lint warning

pruthvihebbani replied:
Fix the warning

pruthvihebbani commented on lte/gateway/c/core/oai/tasks/sgw/spgw_state_manager.cpp:
```diff
@@ -93,13 +96,13 @@ void SpgwStateManager::free_state() {
-        "An error occurred while destroying SGW s11_bearer_context_information "
-        "hashtable");
+  if (state_ue_map.map && state_ue_map.destroy_map() != PROTO_MAP_OK) {
+    OAI_FPRINTF_ERR("An error occurred while destroying SPGW's state ue map ");
```
What doe FPRINTF do? Can we replace this with OAILOG_ERROR

rsarwad replied:
I think we need to keep "OAI_FPRINTF_ERR" because free_state() would be called when tasks exits. So can't rely on logging task as it might have already exited. 
Let me change at other places in this function.

pruthvihebbani commented on lte/gateway/c/core/oai/tasks/sgw_s8/sgw_s8_handlers.cpp:
```diff
@@ -1475,7 +1474,7 @@
     pgw_base_proc_t* base_proc1 = LIST_FIRST(sgw_context_p->pending_procedures);
     LIST_REMOVE(base_proc1, entries);
-    free_wrapper((void**)&sgw_context_p->pending_procedures);
+    free_cpp_wrapper((void**)&sgw_context_p->pending_procedures);
```
Fix the warning

pruthvihebbani commented on lte/gateway/c/core/oai/tasks/sgw/spgw_state.cpp:
```diff
@@ -133,16 +136,22 @@
       p2 = LIST_NEXT(p1, entries);
       LIST_REMOVE(p1, entries);
-      free_wrapper((void**)&p1);
+      free_cpp_wrapper((void**)&p1);
```
Fix the warning

rsarwad committed:
Addressed review comments

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

Commit:
feat(agw): time before killing containers during restart is reduced as (#14139)

otherwise hanging containers take up to 20s to restart and tests run
into timeouts.

Signed-off-by: Marco Pfirrmann <marco.pfirrmann@tngtech.com>
Co-authored-by: Marco Pfirrmann <christian.kraemer@tngtech.com>

Addressed review comment

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

feat(agw): time before killing containers during restart is reduced as (#14139)

otherwise hanging containers take up to 20s to restart and tests run
into timeouts.

Signed-off-by: Marco Pfirrmann <marco.pfirrmann@tngtech.com>
Co-authored-by: Marco Pfirrmann <christian.kraemer@tngtech.com>

Addressed review comment

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

feat(agw): time before killing containers during restart is reduced as (#14139)

otherwise hanging containers take up to 20s to restart and tests run
into timeouts.

Signed-off-by: Marco Pfirrmann <marco.pfirrmann@tngtech.com>
Co-authored-by: Marco Pfirrmann <christian.kraemer@tngtech.com>

fix(ci): there was a bit too much deleted in #14050 (#14143)

Signed-off-by: Fritz Lehnert <13189449+Neudrino@users.noreply.github.com>

rsarwad committed:
Resolved merge conflicts

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into 11190_spgw_ue_context_ht

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

pruthvihebbani commented on lte/gateway/c/core/oai/tasks/sgw/spgw_state_manager.cpp:
```diff
@@ -112,8 +114,7 @@ status_code_e SpgwStateManager::read_ue_state_from_db() {
     OAILOG_DEBUG(log_task, "Reading UE state from db for key %s", key.c_str());
-    spgw_ue_context_t* ue_context_p =
-        (spgw_ue_context_t*)calloc(1, sizeof(spgw_ue_context_t));
+    spgw_ue_context_t* ue_context_p = new spgw_ue_context_t();
```
Null check for ue_context_p

ajahl replied:
In C++  the `new` operator already throws an exception (`bad_alloc`) if it is not able to allocate memory I guess. Why we should need the Null check?

rsarwad replied:
@ajahl should we catch the exception?

ajahl replied:
Only if we can handle the error, maybe to log the error. However, the application should be stopped.

rsarwad replied:
Then it's good to catch the exception to understand where the application is going wrong. @ssanadhya May I know your thoughts on this?

ssanadhya replied:
@rsarwad , since this would be a system error, it is best to let the process terminate as a result of `bad_alloc` exception. No need to catch that exception.

pruthvihebbani commented on lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.cpp:
```diff
@@ -55,8 +55,7 @@ status_code_e mock_read_spgw_ue_state_db(
 
-    spgw_ue_context_t* ue_context_p =
-        (spgw_ue_context_t*)calloc(1, sizeof(spgw_ue_context_t));
+    spgw_ue_context_t* ue_context_p = new spgw_ue_context_t();
```
Null check missing

pruthvihebbani commented on lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:
```diff
@@ -76,8 +76,7 @@ TEST(SPGWStateConverterTest, TestUEContextConversion) {
     oai::SpgwUeContext proto_ctx;
-    spgw_ue_context_t* final_ctx =
-        (spgw_ue_context_t*)calloc(1, sizeof(spgw_ue_context_t));
+    spgw_ue_context_t* final_ctx = new spgw_ue_context_t();
```
Null check missing

pruthvihebbani commented on lte/gateway/c/core/oai/tasks/sgw/spgw_state_converter.cpp:
```diff
@@ -937,26 +936,23 @@ void SpgwStateConverter::insert_proc_into_sgw_pending_procedures(
   for (auto& eps_bearer_proto : proto.pending_eps_bearers()) {
-    sgw_eps_bearer_ctxt_t* eps_bearer =
-        (sgw_eps_bearer_ctxt_t*)calloc(1, sizeof(sgw_eps_bearer_ctxt_t));
+    sgw_eps_bearer_ctxt_t* eps_bearer = new sgw_eps_bearer_ctxt_t();
```
NULL check missing

pruthvihebbani commented on lte/gateway/c/core/oai/tasks/sgw/spgw_state_converter.cpp:
```diff
@@ -937,26 +936,23 @@ void SpgwStateConverter::insert_proc_into_sgw_pending_procedures(
         pending_procedures) {
-  pgw_ni_cbr_proc_t* s11_proc_create_bearer =
-      (pgw_ni_cbr_proc_t*)calloc(1, sizeof(pgw_ni_cbr_proc_t));
+  pgw_ni_cbr_proc_t* s11_proc_create_bearer = new pgw_ni_cbr_proc_t();
```
Null check missing

pruthvihebbani commented on lte/gateway/c/core/oai/tasks/sgw/spgw_state_converter.cpp:
```diff
@@ -233,8 +233,7 @@ void SpgwStateConverter::proto_to_sgw_pdn_connection(
     if (proto.eps_bearer_list(i).eps_bearer_id()) {
-      auto* eps_bearer_entry =
-          (sgw_eps_bearer_ctxt_t*)calloc(1, sizeof(sgw_eps_bearer_ctxt_t));
+      auto* eps_bearer_entry = new sgw_eps_bearer_ctxt_t();
```
Null check missing

pruthvihebbani commented on lte/gateway/c/core/oai/tasks/sgw/pgw_procedures.cpp:
```diff
@@ -72,8 +70,7 @@ pgw_ni_cbr_proc_t* pgw_create_procedure_create_bearer(
   s11_proc_create_bearer->pending_eps_bearers =
-      (pgw_ni_cbr_proc_s::pending_eps_bearers_s*)calloc(
-          1, sizeof(s11_proc_create_bearer->pending_eps_bearers));
+      new pgw_ni_cbr_proc_s::pending_eps_bearers_s();
```
Null check missing

pruthvihebbani commented on lte/gateway/c/core/oai/tasks/sgw/pgw_procedures.cpp:
```diff
@@ -54,9 +54,7 @@ void delete_pending_procedures(
-  pgw_ni_cbr_proc_t* s11_proc_create_bearer =
-      reinterpret_cast<pgw_ni_cbr_proc_t*>(
-          calloc(1, sizeof(pgw_ni_cbr_proc_t)));
+  pgw_ni_cbr_proc_t* s11_proc_create_bearer = new pgw_ni_cbr_proc_t();
```
Null check missing

rsarwad committed:
Addressed review comment

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

ajahl commented on lte/gateway/c/core/oai/tasks/sgw/spgw_state_manager.cpp:
```diff
@@ -58,9 +58,13 @@ void SpgwStateManager::create_state() {
-      hashtable_ts_create(SGW_STATE_CONTEXT_HT_MAX_SIZE, nullptr,
-                          (void (*)(void**))sgw_free_ue_context, nullptr);
+  state_ue_map.map = new google::protobuf::Map<uint64_t, spgw_ue_context_s*>();
+  if (!(state_ue_map.map)) {
```
If you use new to instantiate an object in C++ the Null check is always needless because of the bad_alloc exception.

ajahl commented on lte/gateway/c/core/oai/tasks/sgw/pgw_handlers.cpp:
```diff
@@ -595,9 +594,7 @@ status_code_e create_temporary_dedicated_bearer_context(
-          calloc(1, sizeof(sgw_eps_bearer_ctxt_t)));
+  sgw_eps_bearer_ctxt_t* eps_bearer_ctxt_p = new sgw_eps_bearer_ctxt_t();
 
   if (!eps_bearer_ctxt_p) {
```
The Null check is always needless because of C++ throws a bad_alloc exception.

ajahl commented on lte/gateway/c/core/oai/tasks/sgw/pgw_handlers.cpp:
```diff
@@ -660,8 +657,7 @@ status_code_e create_temporary_dedicated_bearer_context(
-      reinterpret_cast<sgw_eps_bearer_entry_wrapper_s*>(
-          calloc(1, sizeof(*sgw_eps_bearer_entry_p)));
+      new sgw_eps_bearer_entry_wrapper_s();
   if (!sgw_eps_bearer_entry_p) {
```
The Null check is always needless because of C++ throws a bad_alloc exception.

ajahl commented on lte/gateway/c/core/oai/tasks/sgw/pgw_procedures.cpp:
```diff
@@ -54,16 +54,25 @@ void delete_pending_procedures(
-      reinterpret_cast<pgw_ni_cbr_proc_t*>(
-          calloc(1, sizeof(pgw_ni_cbr_proc_t)));
+  pgw_ni_cbr_proc_t* s11_proc_create_bearer = new pgw_ni_cbr_proc_t();
+  if (!s11_proc_create_bearer) {
```
The Null check is always needless because of C++ throws a bad_alloc exception.

ajahl commented on lte/gateway/c/core/oai/tasks/sgw/pgw_procedures.cpp:
```diff
@@ -54,16 +54,25 @@ void delete_pending_procedures(
   if (!ctx_p->pending_procedures) {
     ctx_p->pending_procedures =
         new sgw_eps_bearer_context_information_s::pending_procedures_s();
+    if (!ctx_p->pending_procedures) {
```
The Null check is always needless because of C++ throws a bad_alloc exception.

ajahl commented on lte/gateway/c/core/oai/tasks/sgw/sgw_context_manager.cpp:
```diff
@@ -116,19 +116,15 @@ mme_sgw_tunnel_t* sgw_cm_create_s11_tunnel(teid_t remote_teid,
-     * Malloc failed, may be ENOMEM error
-     */
+  new_tunnel = new mme_sgw_tunnel_t();
+  if (new_tunnel == nullptr) {
```
The Null check after call the new operator is always needless because of C++ throws a bad_alloc exception.

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into 11190_spgw_ue_context_ht

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Catching exception on failing to allocate memory

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Removed the null check

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

pruthvihebbani reviewed (approved):
LGTM

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into 11190_spgw_ue_context_ht

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>
